### PR TITLE
158 change ssot metadata

### LINKED ISSUE #158: change SSOT to have metadata on top level
As a developer, I'd like to have all information related to the robot on the top level of the document. Therefore the SSOT model and the existing documents should be changed to reflect those changes.

---

### :thought_balloon: &nbsp; Hints & Thoughts

- Please also add the attribute of the robotName
- the attributes of robotName, starterId should be at the top level of the document
- there should no longer be a robotId field on the SSoT document/model; as we are using MongoDB's Id, we can reference the SSoT by using the _id field which is automatically generated at the top level

---

## :white_check_mark: &nbsp; Conditions of satisfaction

- [x] Mongoose models have been updated
- [x] backend routes related to SSoT keep existing functionality and return the same values as before
- [x] existing models in the MongoDB have been changed
- [x] the wiki has been updated to represent the changes

### PR DESCRIPTION
**Pull request by WolfgangDaniel**
closes #158 

Don't get confused by the name of the branch. It should be 158 instead of 105.

## Changes
We decided to update the ssot structure by deleting the robotMetadata tag and to have four high level tags from now on:  _id, starterId, robotName, elements. In order to achieve that I changed the existing ssot in the database, I changed the mongoose ssot model and I changed all the files that were either referring to robotMetadata or robotId and needed to be changed.

## Checklist before merge

**Developer's responsibilities**
* [ ] **Assign** one or two developers
* [ ] **Change code** if reviewer(s) has/have requested it
* [ ] **Pull request build** has passed
* [ ] **tested locally** (in at least chrome & firefox if frontend)
* [ ] updated the **documentation**
* [ ] added **tests** where necessary

### CODE REVIEW

File: client/src/components/pages/RobotOverview/RobotOverview.jsx [hunk 1]
```diff
@@ -67,18 +67,13 @@ const RobotOverview = () => {
...
     );
 
     return (
       <>
         {filteredBotList.map((val) => (
-          <RobotContainer 
-            robotId={val.robotId} 
-            robotName={val.robotName} 
-          />
+          <RobotContainer robotId={val['_id']} robotName={val.robotName} />
```
Comment (lukashueller) [inline client/src/components/pages/RobotOverview/RobotOverview.jsx:76]: ESLINT says, that you should use the dot notation here `robotId={val.id}`

File: server/controllers/ssotRetrievalController.js [hunk 2]
```diff
@@ -5,75 +5,92 @@ const userAccessModels = require('../models/userAccessObjectModel.js');
...
-        ).exec();
-        
-        const SSOTs = [];
-        availableSSOTs.forEach((singleSSOT) => {
-            SSOTs.push(singleSSOT.robotMetadata);
-        });
+    const SSOTs = [];
+    availableSSOTs.forEach((singleSSOT) => {
+      SSOTs.push({
+        _id: singleSSOT['_id'],
```
Comment (lukashueller) [inline server/controllers/ssotRetrievalController.js:57]: Same thing here as above. If this is really as simple as I assume, please fix it.

**Review by lukashueller — COMMENTED** [review]
Thank you for addressing the issue. So early in the sprint it certainly makes sense to make such a change. I only noted minor things in two lines that throw an ESLINT error. 

_However, one thing catches my eye in general:_ The large number of changed lines that only came about because of different Prettier configurations. We should really use a uniform configuration there asap, that adjustments like `"testString"` => `'testString'` do not pollute our changed files.

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 1 above)
Comment (kej-jay) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:76]: @WolfgangDaniel the field is called _id in MongoDB but it is possible to access that field through calling it by just id. I learned this just recently, but that being said, the linter is probably right :)

**Review by kej-jay — COMMENTED** [review]
Looks good in most parts. Please consider the small adjustments I marked.

File: client/src/components/pages/RobotOverview/RobotOverview.jsx [hunk 3]
```diff
@@ -67,18 +67,13 @@ const RobotOverview = () => {
    */
   const createRobotBoxes = (searchValue2) => {
```
[the remark this answers is not in the document]
Comment (SanJSp) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:68]: https://github.com/bptlab/ark_automate/pull/138#discussion_r589317578 was supposed to be resolved :D maybe Daniel can change this now :D

File: server/controllers/ssotRetrievalController.js [hunk 4]
```diff
@@ -5,75 +5,92 @@ const userAccessModels = require('../models/userAccessObjectModel.js');
...
+        }
+      )
+      .exec();
 
-        res.send(ssot.robotMetadata);
-    } catch (err) {
-        console.error(err);
-    }
+    res.send({
+      starterId: ssot.startedId,
```
[the remark this answers is not in the document]
Comment (SanJSp) [reply server/controllers/ssotRetrievalController.js:90]: We should rethink the use of starterId in general. We can generate the starter Element by checking the predecessors of each element and the one without any is the starterId.

File: server/controllers/ssotRetrievalController.js [hunk 5]
```diff
@@ -5,75 +5,92 @@ const userAccessModels = require('../models/userAccessObjectModel.js');
...
-            }
-        ).exec();
-        
-        const SSOTs = [];
-        availableSSOTs.forEach((singleSSOT) => {
-            SSOTs.push(singleSSOT.robotMetadata);
-        });
+    const SSOTs = [];
+    availableSSOTs.forEach((singleSSOT) => {
+      SSOTs.push({
```
Comment (SanJSp) [inline server/controllers/ssotRetrievalController.js:56]: camel case

File: server/controllers/ssotRetrievalController.js [hunk 6]
```diff
@@ -5,75 +5,92 @@ const userAccessModels = require('../models/userAccessObjectModel.js');
...
-                robotMetadata: 1
-            }
-        ).exec();
-        
-        const SSOTs = [];
-        availableSSOTs.forEach((singleSSOT) => {
-            SSOTs.push(singleSSOT.robotMetadata);
-        });
+    const SSOTs = [];
+    availableSSOTs.forEach((singleSSOT) => {
```
Comment (SanJSp) [inline server/controllers/ssotRetrievalController.js:55]: single Single Source Of Truth is a bit weird - maybe use `individualSsot`?

File: server/services/BpmnToSsotParsing/BpmnToSsotParsing.js [hunk 7]
```diff
@@ -15,117 +15,127 @@ const ssotMetadataObj = SsotBaseObjects.metadataObj;
...
-    const bpmnActivities = bpmnJson['bpmn2:definitions']['bpmn2:process']['bpmn2:task'];
-    elementsArray = enrichInstructionElements(elementsArray, bpmnActivities)
-    elementsArray = enrichMarkerElements(elementsArray)
-
-    ssot.elements = elementsArray
-    return JSON.stringify(ssot, null, 2);
+  const ssot = {};
+  ssot['_id'] = ssotMetadataObj['_id'];
+  ssot.starterId = ssotMetadataObj.starterId;
+  ssot.robotName = ssotMetadataObj.robotName;
```
Comment (SanJSp) [inline server/services/BpmnToSsotParsing/BpmnToSsotParsing.js:127]: please insert these fileds already in the object constructor, like the following:
```javascript
const ssot = {
  _id = ssotMetadataObj['_id'];
  starterId = ssotMetadataObj.starterId;
  robotName = ssotMetadataObj.robotName;
}
```
Additionally, as we do not have a metadataObj anymore, these changes should also be represented by the code. Please move the hardcoded metadata values from `baseObjects` here and delete everything concerning the metadataObject

```javascript
const ssot = {
  _id: '6045eccfa9a07940e5763f0b',
  starterId: 'exampleID',
  robotName: 'exampleRobot',
}
```

**Review by SanJSp — CHANGES_REQUESTED** [review]
uff, those changes were hard to read.
I added some comments. Additionally, I would like to discuss with @kej-jay if we still need the starterId field and whether the removal of this should be also part of this user story.

File: server/services/BpmnToSsotParsing/_tests_/SsotMockForTesting.js [hunk 8]
```diff
@@ -3,61 +3,60 @@ This is a bpmn xml parsed to json. It contains two activites ordered and created
...
+    },
+    {
+      type: 'MARKER',
+      name: '',
+      id: 'Event_0otufrj',
+      predecessorIds: ['Activity_1vb45u8'],
+      successorIds: [],
+    },
+  ],
+};
```
Comment (SanJSp) [inline server/services/BpmnToSsotParsing/_tests_/SsotMockForTesting.js:60]: can you please undo the changes to the quotation marks? This makes the json here unusable for other testing (f.e. exterrnal) Same for json below

**Review by SanJSp — CHANGES_REQUESTED** [review]
Just noticed this while testing

File: server/controllers/ssotRetrievalController.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kej-jay) [reply server/controllers/ssotRetrievalController.js:90]: I don't see a point why we shouldn't save it. It does not take away too much space and could be helpful when working with other modelling interfaces. And in my understanding, it is easier to create the needed order if we already have a start element. It doesn't make it much easier, but we keep the option to do so.

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 1 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:76]: ✅

File: server/controllers/ssotRetrievalController.js (hunk 2 above)
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:57]: ✅

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:68]: ✅ renamed it to currentSearchValue

File: client/src/components/pages/RobotOverview/RobotOverview.jsx [hunk 9]
```diff
@@ -107,10 +102,18 @@ const RobotOverview = () => {
             onChange={changeUserId}
             defaultValue='80625d115100a2ee8d8e695b'
           >
-            <Option key='user1' value='80625d115100a2ee8d8e695b'>Lukas</Option>
-            <Option key='user2' value='365889fcf871cfe88711b630'>Erik</Option>
-            <Option key='user3' value='225aa14130f2c07789cfa38e'>Sandro</Option>
-            <Option key='user4' value='afdb3d839ca02a106750a2a0'>Daniel</Option>
+            <Option key='user1' value='80625d115100a2ee8d8e695b'>
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:105]: I am not changing this on purpose. Prettier is doing it and prettier is on standard configuration in my vscode. I actually don't mind it though.

File: server/controllers/ssotRetrievalController.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:90]: I agree with Erik on this one.

[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:90]: And I changed it to starterId

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:105]: I think we should leave it as is here and discuss in a seperate meeting in general if and how we want to change the prettier settings collectively :)

File: server/controllers/ssotRetrievalController.js [hunk 10]
```diff
@@ -5,75 +5,92 @@ const userAccessModels = require('../models/userAccessObjectModel.js');
...
+          AccessLevel: 0,
+          _id: 0,
+          userId: 0,
+        }
+      )
+      .exec();
+
+    const SSOTIds = [];
+    availableSSOTsById.forEach((singleUserObj) => {
+      SSOTIds.push(singleUserObj.robotId);
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:40]: This was already changed by @lukashueller in PR #175

[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:40]: ✅

File: server/controllers/ssotRetrievalController.js (hunk 5 above)
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:56]: ✅

File: server/controllers/ssotRetrievalController.js (hunk 6 above)
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:55]: ✅ I just changed it to ssot

File: server/services/BpmnToSsotParsing/BpmnToSsotParsing.js (hunk 7 above)
Comment (WolfgangDaniel) [reply server/services/BpmnToSsotParsing/BpmnToSsotParsing.js:127]: ✅

File: server/services/BpmnToSsotParsing/_tests_/SsotMockForTesting.js (hunk 8 above)
Comment (WolfgangDaniel) [reply server/services/BpmnToSsotParsing/_tests_/SsotMockForTesting.js:60]: If this is supposed to be a JSON it is not - it is a JS Object. I changed the files to be json files. Now it should work :)

Comment (WolfgangDaniel) [reply server/services/BpmnToSsotParsing/_tests_/SsotMockForTesting.js:60]: ✅

**Review by kej-jay — APPROVED** [review]
As far as I can tell everything looks good now

**Review by kej-jay — DISMISSED** [review]

**Review by lukashueller — DISMISSED** [review]

**Review by kej-jay — DISMISSED** [review]

**Review by SanJSp — APPROVED** [review]

**Review by kej-jay — APPROVED** [review]

**Review by lukashueller — APPROVED** [review]
